Feat/intf 2482 add table hover (#231)
Kong/kongponents: docs/components/table.md

```
## Props
### Hover

> Highlight the table row on hover

Reviewer: now that highlight is true by default, lets change this example and doc to explain its always on and you can set it to false `hasHover="false"`

  isSmall />
```

### Clickable
Adds `cursor: pointer` and `user-select: none` styling. 

<KTable :options="$frontmatter.tableOptions" isClickable />
```vue
<KTable
  :options="tableOptions"
  isClickable />
```

> ### Custom Row Attributes
> 
> ## hasSidebar
> Adds left border to each table row. By default set to true. The colors can be overridden by themes.
> The below example demostrates the disabled state:
> 
> <template>
>   <KTable
>     :options="tableOptions"
>     :hasSideBorder="false"
>     />
> </template>
> 
> ```vue
> <template>
>   <KTable
>     :options="tableOptions"
>     :hasSideBorder="false"
>     />
> </template>
> ```
> 

Reviewer: ```suggestion
### hasSidebar
Adds left border to each table row. By default set to true. The colors can be overridden by themes.
The below example demonstrates the disabled state:

<template>
  <KTable
    :options="tableOptions"
    :hasSideBorder="false"
    />
</template>

```vue
<template>
  <KTable
    :options="tableOptions"
    :hasSideBorder="false"
    />
</template>
```

## Custom Row Attributes
```

Reviewer: can you copy this formatting to https://github.com/Kong/kongponents/pull/237

          { label: 'Value', key: 'value' },
          { label: 'Enabled', key: 'enabled'}
        ],
        data: [
          {
            type: 'desktop',
            value: 'Windows 10',
            enabled: 'true'
          },
          {
            type: 'phone',
            value: 'LineageOS',
            enabled: 'false'
          },
          {
            type: 'tablet',
            value: 'ipadOS',
            enabled: 'true'
          }
        ]
      }
    }
  },
  methods: {
    rowAttrsFn (rowItem) {
      return {
        class: {
          'enabled': rowItem.enabled === 'true',
          'disabled': rowItem.enabled === 'false'
        },
        'data-testid': 'row-item'
      }
    }
  }
}
</script>
<style>
.k-table {
  tr.enabled {
    --KTableHover: var(--green-200, #ccffe1);
    --KTableBorder: var(--green-400, #19a654);
  }

  tr.disabled {
    --KTableHover: var(--yellow-100, #fff9e6);
    --KTableBorder: var(--yellow-200, #ffdc73);
  }
}
</style>
```

## Events
Bind any DOM [events](https://developer.mozilla.org/en-US/docs/Web/Events) to
various parts of the table.

### Rows
- `@row:<event>` - returns the `Event`, the row item, and the type: `row`

> <template>

Reviewer: This isn't needed. Can you revert to match what currently exists for the events section [here](https://kongponents.netlify.app/components/table.html#events). 
```
## Events
### rows
### cells
`example`
```

  isSmall />
```

### Clickable
Adds `cursor: pointer` and `user-select: none` styling. 

<KTable :options="$frontmatter.tableOptions" isClickable />
```vue
<KTable
  :options="tableOptions"
  isClickable />
```

### hasSidebar
Adds left border to each table row. By default set to true. The colors can be overridden by themes.
The below example demonstrates the disabled state:

<template>
  <KTable
    :options="tableOptions"
    :hasSideBorder="false"
    />
</template>

```vue
<template>
  <KTable
    :options="tableOptions"
    :hasSideBorder="false"
    />
</template>
```

> ## rowAttrs
> Add custom properties to individual rows. The row object is passed as a param.

Reviewer: ```suggestion
## Custom Row Attributes
Add custom properties to individual rows. The row object is passed as a param.

## rowAttrs - Function that returns an object comprising the attributes.
```

  isSmall />
```

### Clickable
Adds `cursor: pointer` and `user-select: none` styling. 

<KTable :options="$frontmatter.tableOptions" isClickable />
```vue
<KTable
  :options="tableOptions"
  isClickable />
```

> ### Custom Row Attributes
> 
> There is a prop called `hasSideBorder` that when passed in, enables a default side border on each row.
> This prop is on by default.
> These colors can be overridden and customized by using theme variables.
> 
> Example with `hasSideBorder` set to false:
> 
> <template>
>   <KTable
>     :options="tableOptions"
>     :hasSideBorder="false"
>     />
> </template>
> 
> ```vue
> <template>
>   <KTable
>     :options="tableOptions"
>     :hasSideBorder="false"
>     />
> </template>
> ```
> 
> Add custom properties to individual rows. The row object is passed as a param.
> 
> - `rowAttrs` - is a Function that returns an object comprising the attributes.
> 
> Example below:
> 
> <template>
>   <KTable
>     :options="tableOptionsRowAttrs"
>     :rowAttrs="rowAttrsFn"
>     />
> </template>

Reviewer: ```suggestion
## hasSidebar
Adds left border to each table row. By default set to true, the below example demos disabling. 

 ```vue
 <KTable
    :options="tableOptions"
    :has-sidebar="false" />
 ``` 

 <KTable
    :options="tableOptions"
    :has-sidebar="false" />

### Custom Row Attributes
Allows the ability to pass additional attributes to each table row.

- `rowAttrs` - is a Function that returns an object comprising the attributes.<template>

export default {
  data() {
    return {
      row: null,
      eventType: '',
      tableOptions: {
        headers: [
          { label: 'Type', key: 'type' },
          { label: 'Value', key: 'value' },
          { label: 'Enabled', key: 'enabled'}
        ],
        data: [
          {
            type: 'desktop',
            value: 'Windows 10',
            enabled: 'true'
          },
          {
            type: 'phone',
            value: 'LineageOS',
            enabled: 'false'
          },
          {
            type: 'tablet',
            value: 'ipadOS',
            enabled: 'true'
          }
        ]
      }
    }
  },
  methods: {
    rowAttrsFn (rowItem) {
      return {
        class: {
          'enabled': rowItem.enabled === 'true',
          'disabled': rowItem.enabled === 'false'
        },
        'data-testid': 'row-item'
      }
    }
  }
}
</script>
<style>
.k-table {
  tr.enabled {
    --KTableHover: var(--green-200, #ccffe1);
    --KTableBorder: var(--green-400, #19a654);
  }

  tr.disabled {
    --KTableHover: var(--yellow-100, #fff9e6);
    --KTableBorder: var(--yellow-200, #ffdc73);
  }
}
</style>
```

> 

Reviewer: Re-add example below

```suggestion
<KTable
    :options="tableOptionsRowAttrs"
    :rowAttrs="rowAttrsFn"
    />
</template>
```

    <div v-else>Waiting</div>

    <KTable 
      :options="tableOptions"
      is-clickable
      @row:click="actionRow" 
      @cell:mouseover="actionRow"
    />
  </div>
</template>

> <template>

Reviewer: this example is no longer showing

Reviewer: Still not seeing this in the docs. Looks like its fixed on the other PR https://github.com/Kong/kongponents/pull/237/files#diff-4975fee17748787a06e99b3259d78e9fR311

  }
}
</script>

> ```

Reviewer: can you add vue to this code block?
![image](https://user-images.githubusercontent.com/5941111/86197564-81cd3d80-bb0a-11ea-9d52-9a32fb05681f.png)

</script>
```

> ### Custom Row Attributes

Reviewer: I think this should be moved out of the "Events" section and be its own before "slots". Also lets not make the border required to pass attrs. I would move `hasSideBorder` up above next to `clickable`,`small`, and `hover` and just say "by default each row has a left border. You can disable by passing false."

Then explain `rowAttrs` is a fn that allows you to bind any attribute to the table row. It passes the row details up so you can conditionally apply attributes. Then show the class example. 
"Here we demonstrate theming the `tableHover` variable based on row data., etc"

Author: question here: `border-spacing: 0 4px` and `border-collapse: separate` should we make this the default css then?

Author: not having any side borders would then turn the border-collapse and border-spacing back to the way it was

that you want the rows to have. `hasSideBorder` prop must be passed in to enable custom row attributes.
In the following example, the rows have a custom green color if enabled is true, and yellow if enabled is false, coming from the
custom `rowAttrs` function. We can also add a custom data-testid onto each row using these attributes (you can inspect it to see there is a custom data-testid attribute with `rowItem` on each row).

<template>
  <KTable
    :options="tableOptions"
    hasSideBorder
    :rowAttrs="rowAttrsFn"
    />
</template>

```vue
<template>
  <KTable
    :options="tableOptions"
    :rowAttrs="rowAttrsFn"
    hasSideBorder
    />
</template>
<script>
import { defaultSorter } from '@kongponents/KTable'
export default {
  data() {
    return {
      row: null,
      eventType: '',
      tableOptions: {
        headers: [
          { label: 'Type', key: 'type' },
          { label: 'Value', key: 'value' }
        ],
        data: [
          {
            name: 'Basic Auth',
            id: '517526354743085',
            enabled: 'true',

>             themeColors: []

Reviewer: lets remove this key, to me its confusing when in an example about theming and its not used to theme

          },
          {
            name: 'Website Desktop',
            id: '328027447731198',
            enabled: 'false',
            themeColors: ['blue','violet']
          },
          {
            name: 'Android App',
            id: '405383051040955',
            enabled: 'true',
            themeColors: ['green', 'yellow']
          }
        ]
      }
    }
  },
  methods: {
    rowAttrsFn (rowItem) {
      return {
        class: {
          'enabled': rowItem.enabled === 'true',
          'disabled': rowItem.enabled === 'false'
        },
        'data-testid': 'row-item'
      }
    }
  }
}
</script>
<style>

> table.k-table.side-border {

Reviewer: here lets just style `.k-table` and not `side-border`

</template>

```vue
<template>
  <KTable
    :options="tableOptions"
    :rowAttrs="rowAttrsFn"
    hasSideBorder
    />
</template>
<script>
import { defaultSorter } from '@kongponents/KTable'
export default {
  data() {
    return {
      row: null,
      eventType: '',
      tableOptions: {
        headers: [
          { label: 'Type', key: 'type' },
          { label: 'Value', key: 'value' }
        ],
        data: [
          {
            name: 'Basic Auth',
            id: '517526354743085',
            enabled: 'true',
            themeColors: []
          },
          {
            name: 'Website Desktop',
            id: '328027447731198',
            enabled: 'false',
            themeColors: ['blue','violet']
          },
          {
            name: 'Android App',
            id: '405383051040955',
            enabled: 'true',
            themeColors: ['green', 'yellow']
          }
        ]
      }
    }
  },
  methods: {
    rowAttrsFn (rowItem) {
      return {
        class: {
          'enabled': rowItem.enabled === 'true',
          'disabled': rowItem.enabled === 'false'
        },
        'data-testid': 'row-item'
      }
    }
  }
}
</script>
<style>

> table.k-table {

Reviewer: This is confusing. If you can theme by css var why demo all this css overrides?

Author: the theming is only overridable for the default border color and hover color but in this case we want to have multiple colors for border and hover so I'm confused how would we use multiple colors here with theming?

Reviewer: You're adding a specific class so instead of redefining every css trait, just set what the var is for each class.

```scss
.k-table
  &.enabled {
    --KTableBorder: var(--green-400);
    --KTableHover: var(--green-100);
  }
    &.disabled {
    --KTableBorder: var(--yellow-400);
    --KTableHover: var(--yellow-100);
  }
}
```

</script>
```

## Custom Row Attributes

> There is a prop called `hasSideBorder` that when passed in, enables a default side border on each row, using the color `steal-300` by default.

Reviewer: Please move this to the top props section https://github.com/Kong/kongponents/pull/231/files#r449151214

Reviewer: I also don't think you need to call out the default color. Just say its customizable with a theme var

This prop is on by default.
These colors can be overridden by themes or by using custom row attributes.
Example with `hasSideBorder` set to false:

<template>
  <KTable
    :options="tableOptions"
    :hasSideBorder="false"
    />
</template>

```vue
<template>
  <KTable
    :options="tableOptions"
    :hasSideBorder="false"
    />
</template>
```

> We can also add custom row attributes via a prop called `rowAttrs` - this prop is a Function that returns an object comprising the conditional custom attributes

Reviewer: Lets move this description up (below "Custom Row Attrs") and update the description to not be a continuation.

```suggestion
Add custom properties to individual rows. The row object is passed as a param.

- `rowAttrs` - returns an object comprising the attributes
```

that you want the rows to have.
In the following example, the rows have a custom green color if enabled is true, and yellow if enabled is false.
We can also add a custom data-testid onto each row using these attributes (you can inspect it to see there is a custom data-testid attribute with `rowItem` on each row).

<template>
  <KTable
    :options="tableOptionsRowAttrs"
    hasSideBorder
    :rowAttrs="rowAttrsFn"
    />
</template>

```vue
<template>
  <KTable
    :options="tableOptions"
    :rowAttrs="rowAttrsFn"

>     hasSideBorder

Reviewer: ```suggestion

```

  .table-wrapper {
  --KTableHover: lavender;
  }

>   table.k-table.side-border {

Reviewer: same as above. `.k-table tr.enabled/.disabled`
